src/wallet/wallet.go

```diff
@@ -96,40 +119,227 @@ func NewWallet(wltName string, opts Options) (*Wallet, error) {
+		return nil, fmt.Errorf("generate addresses failed when creating wallets: %v", err)
+	}
+
+	if !opts.Encrypt {
+		return w, nil
+	}
+
+	if err := w.lock(opts.Password); err != nil {
+		return nil, fmt.Errorf("lock wallet failed: %v", err)
+	}
+
+	// Update the encrypted meta field
+	w.setEncrypted(true)
+
 	return w, nil
 }
 
+// lock encrypts the wallet with password
+func (w *Wallet) lock(password []byte) error {
+	if len(password) == 0 {
+		return ErrRequirePassword
+	}
+
+	if w.IsEncrypted() {
+		return ErrWalletEncrypted
+	}
+
+	// Encrypt the seed
+	ss, err := Encrypt([]byte(w.seed()), password)
+	if err != nil {
+		return err
+	}
+
+	w.setSeed(ss)
+
+	// Encrypt the last seed
+	sls, err := Encrypt([]byte(w.lastSeed()), password)
+	if err != nil {
+		return err
+	}
+
+	w.setLastSeed(sls)
+
+	// encrypt private keys in entries
+	for i, e := range w.Entries {
+		se, err := Encrypt(e.Secret[:], password)
+		if err != nil {
+			return err
+		}
+
+		// Set the encrypted seckey value
+		w.Entries[i].EncryptedSecret = se
+		// Clear the entry.Secret
+		for j := range w.Entries[i].Secret {
+			w.Entries[i].Secret[j] = 0
+		}
+
+		w.Entries[i].Secret = cipher.SecKey{}
+	}
+
+	w.setEncrypted(true)
+
+	return nil
+}
+
+// unlock decrypts the wallet into a temporary decrypted copy of the wallet
+// It returns an error if decryption fails
+// The temporary decrypted wallet should be erased from memory when done.
+func (w *Wallet) unlock(password []byte) (*Wallet, error) {
+	if !w.IsEncrypted() {
+		return nil, ErrWalletNotEncrypted
+	}
+
+	if len(password) == 0 {
+		return nil, errors.New("password is required to decrypt wallet")
+	}
+
+	wlt := w.clone()
+
+	// decrypt the seed
+	s, err := Decrypt(wlt.seed(), password)
+	if err != nil {
+		return nil, err
+	}
+	wlt.setSeed(string(s))
+
+	// decrypt lastSeed
+	ls, err := Decrypt(wlt.lastSeed(), password)
+	if err != nil {
+		return nil, err
+	}
+	wlt.setLastSeed(string(ls))
+
+	// decrypt the entries
+	for i := range wlt.Entries {
+		sk, err := Decrypt(wlt.Entries[i].EncryptedSecret, password)
+		if err != nil {
+			return nil, err
+		}
+		copy(wlt.Entries[i].Secret[:], sk[:])
+		wlt.Entries[i].EncryptedSecret = ""
+	}
+	wlt.setEncrypted(false)
+
+	return wlt, nil
+}
+
+// guard will do:
+// 1. unlock the encrypted wallet
+// 2. process the decrypted wallet by calling the callback function
+// 3. lock the wallet at the end
+// If the wallet is not encrypted, it would return ErrWalletNotEncrypted error
+func (w *Wallet) guard(password []byte, f func(w *Wallet) error) (err error) {
+	if !w.IsEncrypted() {
+		return ErrWalletNotEncrypted
+	}
+
+	if len(password) == 0 {
+		return ErrRequirePassword
+	}
+
+	var wlt *Wallet
+	wlt, err = w.unlock(password)
+	if err != nil {
+		return fmt.Errorf("unlock wallet failed: %v", err)
+	}
+
+	defer func() {
+		if lockErr := wlt.lock(password); lockErr != nil {
+			err = fmt.Errorf("lock wallet failed: %v", lockErr)
+			return
+		}
+
+		*w = *wlt
```

Comment:
Change this to do:

if `f(wlt)` succeeds, then lock(wlt), copy to w.  lock(wlt) should return a copy of wlt, if not, then clone wlt before copy to w.

Add an erase method to wipe secret fields.  In the defer, call erase() on wlt

Reply:
We use the `seed` field to store either raw seed or encrypted seed. Once the `unlock` method is called, the encrypted seed value will be replaced with decrypted seed value, so if `f(wlt)` failed, calling `erase()` could only wipe the secrets, can't set back the original encrypted seed value, so does the `lastSeed`.

To implement this, I will add `encryptedSeed` and `encryptedLastSeed` fields, call `unlock` method to decrypt them and set the decrypted value in `seed` and `lastSeed` fields, call `erase` method to wipe the `seed` and `lastSeed` value.